src/marker.rs

```diff
@@ -522,26 +337,282 @@ impl<'py> Python<'py> {
+    ///     py.safe_allow_threads(move || {
     ///         println!("{:?}", s); // This causes a compile error.
     ///     });
     /// }
     /// ```
     ///
+    /// # Example: The `send_wrapper` loophole is closed by running the closure on dedicated thread
+    ///
+    /// ```should_panic
+    /// # use pyo3::prelude::*;
+    /// # use pyo3::types::PyString;
+    /// use send_wrapper::SendWrapper;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let string = PyString::new(py, "foo");
+    ///
+    ///     let wrapped = SendWrapper::new(string);
+    ///
+    ///     py.safe_allow_threads(|| {
+    ///         // panicks because this is not the thread which created `wrapped`
+    ///         let sneaky: &PyString = *wrapped;
+    ///         println!("{:?}", sneaky);
+    ///     });
+    /// });
+    /// ```
+    ///
     /// [`Py`]: crate::Py
     /// [`PyString`]: crate::types::PyString
     /// [auto-traits]: https://doc.rust-lang.org/nightly/unstable-book/language-features/auto-traits.html
     /// [Parallelism]: https://pyo3.rs/main/parallelism.html
-    pub fn allow_threads<T, F>(self, f: F) -> T
+    pub fn safe_allow_threads<T, F>(self, f: F) -> T
+    where
+        F: Send + FnOnce() -> T,
+        T: Send,
+    {
+        use std::mem::{replace, transmute};
+        use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+        use std::sync::Arc;
+        use std::thread::Builder;
+        use std::time::Duration;
+
+        use parking_lot::{const_mutex, Condvar, Mutex};
+
+        use crate::impl_::panic::PanicTrap;
+
+        // Use a guard pattern to handle reacquiring the GIL,
+        // so that the GIL will be reacquired even if `f` panics.
+        // The `Send` bound on the closure prevents the user from
+        // transferring the `Python` token into the closure.
+        let _guard = unsafe { SuspendGIL::new() };
+
+        // To close soundness loopholes w.r.t. `send_wrapper` or `scoped-tls`,
+        // we run the closure on a separate thread so that it cannot
+        // access thread-local storage from the current thread.
+
+        // 1. Construct a task
+        struct Task(*mut (dyn FnMut() + 'static));
+        unsafe impl Send for Task {}
+
+        let mut f = Some(f);
+        let mut result = None;
+
+        let mut task = || {
+            let f = f
+                .take()
+                .expect("allow_threads closure called more than once");
+
+            result = Some(catch_unwind(AssertUnwindSafe(f)));
+        };
+
+        // SAFETY: the current thread will block until the closure has returned
+        let mut task = Task(unsafe { transmute(&mut task as &mut (dyn FnMut() + '_)) });
+
+        // 2. Dispatch task to waiting thread, spawn new thread if necessary
+        let trap = PanicTrap::new(
+            "allow_threads panicked while stack data was accessed by another thread which is a bug",
+        );
+
+        enum MailboxInner {
```

Comment:
This implementation looks really solid. I kept wondering about small micro-optimisations like replacing the `Arc` with a `Box` and a raw pointer, but I felt they were likely all irrelevant compared the the initial cost of creating the thread. The warm path with an idle thread already ready to go looks about as fast as it can possibly be.